torch/_refs/__init__.py

```diff
@@ -3689,63 +3702,140 @@ def tril(a: TensorLikeType, diagonal: int = 0) -> TensorLikeType:
+    n_row_trapezoid = m_last_row - m_first_row + 1
+
+    # Number of elements in top trapezoid
+    trapezoid_size = (m_first_row + m_last_row) * n_row_trapezoid // 2
+    # Number of elemetns in bottom rectangle
+    diff_row = n_row_all - n_row_trapezoid
+    rectangle_size = max(0, (n_row_all - n_row_trapezoid) * col)
+
+    return trapezoid_size, rectangle_size, m_first_row
+
+
+# This is based on tril_indices_cuda in aten/src/ATen/native/cuda/TensorFactories.cu
+@register_decomposition(torch.ops.aten.tril_indices)
+def tril_indices(
     row: int,
     col: int,
     offset: int = 0,
     *,
-    dtype: Optional[torch.dtype] = None,
-    layout: Optional[torch.layout] = None,
-    device: Optional[torch.device] = None,
+    dtype: torch.dtype = torch.long,
+    layout: torch.layout = torch.strided,
+    device: DeviceLikeType = "cpu",
     pin_memory: bool = False,
 ) -> TensorLikeType:
     if pin_memory:
         raise NotImplementedError("PrimTorch doesn't support pinned memory")
     check(row >= 0, lambda: f"row must be non-negative, got {row}")
     check(col >= 0, lambda: f"row must be non-negative, got {col}")
-    # We need two versions of the xs,ys: one with long dtype and one requested dtype
-    # The long dtype is to compute the mask, so it doesn't get messed up due to overflow in small dtypes
-    xs_long = torch.arange(row, device=device).unsqueeze(1)
-    ys_long = torch.arange(col, device=device).unsqueeze(0)
-    xs = torch.arange(row, dtype=dtype, layout=layout, device=device).unsqueeze(1)
-    ys = torch.arange(col, dtype=dtype, layout=layout, device=device).unsqueeze(0)
-    mask = ys_long >= xs_long + offset
-    return torch.stack((xs.masked_select(mask), ys.masked_select(mask)))
+    check(
+        layout == torch.strided,
+        lambda: f"only support layout=torch.strided, got {layout}",
+    )
+
+    trapezoid_size, rectangle_size, m_first_row = _get_tril_sizes(row, col, offset)
+    row_offset = max(0, -offset)
+
+    # first we do the indices for top trapezoid
+    xs1 = torch.arange(0, trapezoid_size, dtype=torch.float64, device=device)
+    b = m_first_row - 0.5
+    row_inds1 = torch.floor(-b + torch.sqrt(b * b + 2 * xs1))
+    col_inds1 = torch.floor(xs1 - (2 * m_first_row - 1 + row_inds1) * row_inds1 * 0.5)
+    row_inds1 = prims.to_dtype(row_inds1 + row_offset, dtype)
+    col_inds1 = prims.to_dtype(col_inds1, dtype)
+
+    # then bottom rectangle
+    xs2 = torch.arange(0, rectangle_size, device=device)
+    row_inds2 = xs2 // col + (col - m_first_row + 1 + row_offset)
+    col_inds2 = xs2 % col
+    row_inds2 = prims.to_dtype(row_inds2, dtype)
+    col_inds2 = prims.to_dtype(col_inds2, dtype)
+
+    return torch.stack(
+        (torch.cat((row_inds1, row_inds2)), torch.cat((col_inds1, col_inds2)))
+    )
 
 
-@register_decomposition(torch.ops.aten.tril_indices)
-def tril_indices(
+def _get_triu_sizes(row: int, col: int, offset: int) -> Tuple[int, int, int]:
```

Comment:
I haven't thought too much about it, but intuitively, isn't `triu_sizes` the same as `tril_sizes` but swapping the row and the column?

Reply:
Nevermind, I see that you use `tril_sizes` tin the implementation here.

Reply:
Actually, yeah, can't this function be implemented as:
```python
return _get_tril_sizes(col, row, -offset)
```
?

Reply:
No, unfortunately it doesn't work that way.

The total size (trapezoid+rectangle) is the same, but the split into trapezoid and rectangle is different.
What happens is that you can split the shape of the lower (upper) indices into a trapezoid and either a top (bottom) rectangle or a left (right) rectangle. For the purposes of the implementation you need the size of the top/bottom rectangle and when you flip rows and columns like that you map bottom rectangle into right rectangle, not the top rectangle that you need.

E.g.,:
```
> torch._refs._get_triu_sizes(2,3,-1)
(3, 3, 3)
> torch._refs._get_tril_sizes(3,2,1)
(2, 4, 2)
```

Reply:
I knew I was missing something. Could you add comments in these two functions (or one of them and reference it from the other) documenting what is it that they do exactly? Thank you!